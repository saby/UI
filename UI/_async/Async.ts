/// <amd-module name='UI/Base:Async' />
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import * as Library from 'WasabyLoader/Library';
import { IoC, constants } from 'Env/Env';
import { descriptor } from 'Types/entity';
import { default as Control, IControlOptions, TemplateFunction } from 'UI/_base/Control';
import { headDataStore } from 'UI/_base/HeadData';
import template = require('wml!UI/_async/Async');

function generateErrorMsg(templateName: string, msg?: string): string {
   const tTemplate = `Ошибка загрузки контрола "${templateName}"`;
   const tHint = 'Возможны следующие причины:\n\t \
                  • Ошибка в самом контроле\n\t \
                  • Долго отвечал БЛ метод в _beforeUpdate\n\t \
                  • Контрола не существует';
   return !msg ? `${tTemplate}\n${tHint}` : `${tTemplate}: ${msg}`;
}

export type TAsyncStateReceived = boolean | string;

export interface IAsyncOptions extends IControlOptions {
   templateName: string;
   templateOptions: IControlOptions;
}

const SUCCESS_BUILDED = 's';

/**
 * Контейнер для асинхронной загрузки контролов.
 * Подробное описание и примеры вы можете найти <a href='/doc/platform/developmentapl/interface-development/pattern-and-practice/async-load/'>здесь</a>.
 *
 * @class UI/Base:Async
 * @extends Core/Control
 *
 * @private
 * @author Санников К.А.
 */
export default abstract class Async extends Control<IAsyncOptions, TAsyncStateReceived> {
   protected _template: TemplateFunction = template;
   protected currentTemplateName: string;
   protected optionsForComponent: Record<string, unknown> = {};
   /**
    * Флаг для того, чтобы избежать повторной загрузки шаблона, при изменении опций до окончания асинхронной загрузки
    */
   private asyncLoading: boolean = false;
   /**
    * Флаг, о том, что произошла ошибка при загрузке модуля - чтобы не было циклической попытки загрузки
    */
   private loadingErrorOccurred: boolean = false;
   protected error: TAsyncStateReceived | void;
   protected userErrorMessage: string | void;
   protected defaultErrorMessage: string = 'У СБИС возникла проблема';

   protected _beforeMount(options: IAsyncOptions, _: unknown, receivedState: TAsyncStateReceived): Promise<TAsyncStateReceived> {
      if (!options.templateName) {
         this.error = 'В модуль Async передали не корректное имя шаблона (templateName=undefined|null|empty)';
         IoC.resolve('ILogger').error(this.error);
         return Promise.resolve(this.error);
      }

      if (receivedState && receivedState !== SUCCESS_BUILDED) {
         IoC.resolve('ILogger').error(receivedState);
      }

      if (constants.isBrowserPlatform && (!ModulesLoader.isLoaded(options.templateName) ||
         this._isCompat() || !receivedState)) {
         return this._loadContentAsync(options.templateName, options.templateOptions);
      }

      this.error = this._loadContentSync(options.templateName, options.templateOptions);
      if (this.error) {
         this.userErrorMessage = this.defaultErrorMessage;
         return Promise.resolve(this.error);
      }

      return Promise.resolve(SUCCESS_BUILDED);
   }

   /**
    * Если можем подставить данные при изменении синхронно, то делаем это до обновления
    * @param {*} opts
    */
   protected _beforeUpdate(opts: IAsyncOptions): void {
      if (this.asyncLoading) {
         return;
      }

      if (opts.templateName === this.currentTemplateName) {
         // поменялись только опции шаблона
         this._insertComponent(this.optionsForComponent.resolvedTemplate,
            opts.templateOptions,
            opts.templateName);
         return;
      }

      if (ModulesLoader.isLoaded(opts.templateName)) {
         this._loadContentSync(opts.templateName, opts.templateOptions);
      }
   }

   /**
    * Если до обновления не загрузили синхронно, то пробуем загрузить асинхронно
    */
   protected _afterUpdate(): void {
      if (this.asyncLoading) {
         return;
      }
      if (this.loadingErrorOccurred) {
         this.loadingErrorOccurred = false;
         return;
      }
      if (this.currentTemplateName === this._options.templateName) {
         return;
      }

      this._loadContentAsync(this._options.templateName, this._options.templateOptions).then(() => {
         this._forceUpdate();
      });
   }

   protected _loadContentSync(name: string, options: IControlOptions): TAsyncStateReceived {
      const loaded = this._loadSync(name);
      if (loaded === null) {
         return generateErrorMsg(name);
      }

      this._insertComponent(loaded, options, name);
      this._pushDepToHeadData(Library.parse(name).name);
      return false;
   }

   protected _loadSync<T = unknown>(name: string): T {
      try {
         const loaded = ModulesLoader.loadSync<T>(name);
         if (loaded) {
            return loaded;
         }
      } catch (err) {
         IoC.resolve('ILogger').error(`Couldn't load module "${name}"`, err);
      }
      return null;
   }

   protected _loadContentAsync(name: string, options: IControlOptions): Promise<TAsyncStateReceived> {
      // Need this flag to prevent setting new options for content
      // that wasn't loaded yet
      this.asyncLoading = true;
      this.loadingErrorOccurred = false;

      return this._loadAsync(name).then<TAsyncStateReceived, TAsyncStateReceived>((loaded) => {
         this.asyncLoading = false;
         if (!loaded) {
            this.loadingErrorOccurred = true;
            this.error = generateErrorMsg(name);
            IoC.resolve('ILogger').warn(this.error);
            this.userErrorMessage = this.defaultErrorMessage;
            return this.error;
         }

         this._insertComponent(loaded, options, name);
         return true;
      }, (err) => {
         this.asyncLoading = false;
         this.loadingErrorOccurred = true;
         this.error = generateErrorMsg(name);
         this.userErrorMessage = err.message;
         return err;
      });
   }

   protected _loadAsync(name: string): Promise<unknown> {
      return ModulesLoader.loadAsync(name).catch((error) => {
         IoC.resolve('ILogger').error(`Couldn't load module "${name}"`, error);
         ModulesLoader.unloadSync(name);
         throw new Error(this.defaultErrorMessage);
      });
   }

   protected _pushDepToHeadData(dep: string): void {
      if (constants.isBrowserPlatform) {
         return;
      }

      try {
         headDataStore.read('pushDepComponent')(dep, true);
      } catch (e) {
         IoC.resolve('ILogger').warn('You\'re trying to use Async without Controls/Application. Link to ' +
            dep +
            ' won\'t be added to server-side generated markup. ' + e);
      }
   }

   protected _insertComponent(tpl: unknown, opts: IControlOptions, templateName: string): void {
      this.error = '';
      this.currentTemplateName = templateName;
      this.optionsForComponent = {};
      for (const key in opts) {
         if (opts.hasOwnProperty(key)) {
            this.optionsForComponent[key] = opts[key];
         }
      }

      if (tpl && tpl['__esModule']) {
         this.optionsForComponent.resolvedTemplate = tpl['default'];
         return;
      }
      this.optionsForComponent.resolvedTemplate = tpl;
   }

   protected _isCompat(): boolean {
      return constants.compat;
   }

   static getOptionTypes(): Record<string, unknown> {
      return {
         templateName: descriptor(String).required()
      };
   }
}

/**
 * @name UI/Base:Async#content
 * @cfg {String} Содержимое контейнера.
 */

/**
 * @name UI/Base:Async#templateName
 * @cfg {String} Имя асинхронно загружаемого контрола.
 * Можно использовать только {@link /doc/platform/developmentapl/interface-development/pattern-and-practice/javascript-libraries/#_2 публичные пути библиотеки}.
 */

/**
 * @name UI/Base:Async#templateOptions
 * @cfg {Object} Параметры содержимого контейнера Async.
 */
