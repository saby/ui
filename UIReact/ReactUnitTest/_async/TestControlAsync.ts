import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!ReactUnitTest/_async/TestControlAsync');

class TestControlAsync extends Control {
    protected _template: TemplateFunction = template;
}

export = TestControlAsync;
