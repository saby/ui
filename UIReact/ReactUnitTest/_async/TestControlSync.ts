import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!ReactUnitTest/_async/TestControlSync');

class TestControlSync extends Control {
    protected _template: TemplateFunction = template;
}

export = TestControlSync;
