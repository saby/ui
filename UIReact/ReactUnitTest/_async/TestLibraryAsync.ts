import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!ReactUnitTest/_async/TestControlAsync');

class ExportControl extends Control {
    protected _template: TemplateFunction = template;
}

function exportFunction(echo: string): string {
    return echo;
}
function exportSyncFunction(echo: string): string {
    return echo;
}

export { exportFunction, exportSyncFunction, ExportControl };
