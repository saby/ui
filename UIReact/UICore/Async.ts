/**
 * Библиотека абстрактной реализации контейнера для асинхронной загрузки контролов.
 * @library UICore/Async
 * @includes Async UICore/_async/Async
 */

export {
    default as Async,
    IAsyncOptions,
    TAsyncStateReceived,
} from 'UICore/_async/Async';
