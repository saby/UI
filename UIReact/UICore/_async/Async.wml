<ws:if data="{{ error }}">
   <div>{{ userErrorMessage }}</div>
</ws:if>
<ws:else data="{{ asyncLoading }}">
   <div>Идет загрузка шаблона...</div>
</ws:else>
<ws:else>
   <ws:partial template="{{ _options.content || optionsForComponent.resolvedTemplate }}" scope="{{ optionsForComponent }}"/>
</ws:else>