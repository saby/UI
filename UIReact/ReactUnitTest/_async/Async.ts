// @ts-nocheck
import { Async, TAsyncStateReceived, IAsyncOptions } from 'UICore/Async';

/**
 * Реализация класса UICore/Async:Async для тестов
 */
export default class AsyncTest extends Async {
    _beforeMount(options: IAsyncOptions): void {
        super._beforeMount(options);
    }

    _componentDidMount(): void {
        super._componentDidMount();
    }

    _beforeUpdate(opts: IAsyncOptions): void {
        super._beforeUpdate(opts);
    }

    _afterUpdate(): void {
        super._afterUpdate();
    }

    getError(): TAsyncStateReceived | void {
        return this.error;
    }

    getCurrentTemplateName(): string {
        return this.currentTemplateName;
    }

    getOptionsForComponent(): Record<string, unknown> {
        return this.optionsForComponent;
    }
}
