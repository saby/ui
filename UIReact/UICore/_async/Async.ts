/**
 * @kaizen_zone 4cbeec99-a5a5-48bf-8531-411cf558b0c0
 */
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import * as Library from 'WasabyLoader/Library';
import { constants } from 'Env/Env';
import { descriptor } from 'Types/entity';
import { Control } from 'UICore/Base';
import { IControlOptions, TemplateFunction } from 'UICommon/Base';
import { addPageDeps } from 'UICommon/Deps';
import { Logger } from 'UICommon/Utils';
import { CommonUtils } from 'UICommon/Executor';
import template = require('wml!UICore/_async/Async');

function generateErrorMsg(templateName: string, msg?: string): string {
    const tTemplate = `Ошибка загрузки контрола "${templateName}"`;
    const tHint =
        'Возможны следующие причины:\n\t \
                  • Ошибка в самом контроле\n\t \
                  • Долго отвечал БЛ метод в _beforeUpdate\n\t \
                  • Контрола не существует';
    return !msg ? `${tTemplate}\n${tHint}` : `${tTemplate}: ${msg}`;
}

export type TAsyncStateReceived = boolean | string;

export interface IAsyncOptions extends IControlOptions {
    templateName: string;
    templateOptions: IControlOptions;
    noBubbleEvents?: boolean;
}

/**
 * Абстрактная реализация контейнера для асинхронной загрузки контролов.
 * !Важно: нельзя использовать этот контейнер напрямую! Необходимо использовать {@link Controls/Container/Async}
 * @see Controls/Container/Async
 * Подробное описание и примеры вы можете найти <a href='/doc/platform/developmentapl/interface-development/pattern-and-practice/async-load/'>здесь</a>.
 *
 * @public
 */
export default abstract class Async extends Control<IAsyncOptions, TAsyncStateReceived> {
    /**
     * @event UICore/Async:Async#templateLoad Событие оповещения, что указанный в templateName шаблон загружен и вставлен в DOM
     */

    protected _template: TemplateFunction = template;
    protected currentTemplateName: string;
    protected optionsForComponent: Record<string, unknown> = {};
    /**
     * Флаг для того, чтобы избежать повторной загрузки шаблона, при изменении опций до окончания асинхронной загрузки
     */
    protected asyncLoading: boolean = false;
    /**
     * Флаг, о том, что произошла ошибка при загрузке модуля - чтобы не было циклической попытки загрузки
     */
    private loadingErrorOccurred: boolean = false;
    protected error: TAsyncStateReceived | void;
    protected userErrorMessage: string | void;
    protected defaultErrorMessage: string = 'У СБИС возникла проблема';
    /**
     * Флаг чтобы понимать, что был загружен контрол и вставлен на страницу -
     * т.к. после монтирования в DOM нужно будет опубликовать событие templateLoad
     * @private
     */
    private needNotifyOnLoad: boolean = false;

    protected _beforeMount(options: IAsyncOptions): Promise<void> | void {
        if (!options.templateName) {
            this.error =
                'В модуль Async передали не корректное имя шаблона (templateName=undefined|null|empty)';
            Logger.error(this.error);
            return;
        }

        if (constants.isBrowserPlatform && !ModulesLoader.isLoaded(options.templateName)) {
            return this._loadContentAsync(options.templateName, options.templateOptions);
        }

        this.error = this._loadContentSync(options.templateName, options.templateOptions);
        if (this.error) {
            this.userErrorMessage = this.defaultErrorMessage;
        }
    }

    protected _componentDidMount(): void {
        this._notifyOnLoad();
    }

    /**
     * Если можем подставить данные при изменении синхронно, то делаем это до обновления
     * @param {*} opts
     */
    protected _beforeUpdate(opts: IAsyncOptions): void {
        if (this.asyncLoading) {
            return;
        }

        if (opts.templateName === this.currentTemplateName) {
            // поменялись только опции шаблона
            this._insertComponent(
                this.optionsForComponent.resolvedTemplate,
                opts.templateOptions,
                opts.templateName
            );
            return;
        }

        if (ModulesLoader.isLoaded(opts.templateName)) {
            this._loadContentSync(opts.templateName, opts.templateOptions);
        }
    }

    _componentDidUpdate(): void {
        if (this.asyncLoading) {
            return;
        }
        if (this.loadingErrorOccurred) {
            this.loadingErrorOccurred = false;
            return;
        }
        if (this.currentTemplateName === this._options.templateName) {
            this._notifyOnLoad();
            return;
        }
        this._loadContentAsync(this._options.templateName, this._options.templateOptions);
    }

    protected _notifyOnLoad(): void {
        if (this.needNotifyOnLoad && !this.error && !this.asyncLoading) {
            this.needNotifyOnLoad = false;
            this._notify('templateLoad');
        }
    }

    protected _loadContentSync(name: string, options: IControlOptions): TAsyncStateReceived {
        const loaded = this._loadSync(name);
        if (loaded === null) {
            return generateErrorMsg(name);
        }

        this.needNotifyOnLoad = true;
        this._insertComponent(loaded, options, name);
        addPageDeps([Library.parse(name).name]);
        return false;
    }

    protected _loadSync<T = unknown>(name: string): T {
        try {
            const loaded = ModulesLoader.loadSync<T>(name);
            if (loaded) {
                return loaded;
            }
        } catch (err) {
            Logger.error(`Couldn't load module "${name}"`, err);
        }
        return null;
    }

    protected _loadContentAsync(name: string, options: IControlOptions): Promise<void> {
        this.asyncLoading = true;
        this.loadingErrorOccurred = false;

        return this._loadAsync(name).then(
            (loaded) => {
                this.asyncLoading = false;
                if (!loaded) {
                    this.loadingErrorOccurred = true;
                    this.error = generateErrorMsg(name);
                    Logger.warn(this.error);
                    this.userErrorMessage = this.defaultErrorMessage;
                    return;
                }

                this.needNotifyOnLoad = true;
                this._insertComponent(loaded, options, name);
            },
            (err) => {
                this.asyncLoading = false;
                this.loadingErrorOccurred = true;
                this._setErrorState(err, name);
            }
        );
    }

    protected _loadAsync(name: string): Promise<unknown> {
        return ModulesLoader.loadAsync(name).catch((error) => {
            Logger.error(`Couldn't load module "${name}"`, error);
            ModulesLoader.unloadSync(name);
            throw new Error(this.defaultErrorMessage);
        });
    }

    protected _setErrorState(err: Error, name: string): void {
        this.error = generateErrorMsg(name);
        this.userErrorMessage = err.message;
    }

    protected _insertComponent(tpl: unknown, options: IControlOptions, templateName: string): void {
        this.error = '';
        this.currentTemplateName = templateName;
        this.optionsForComponent = {};
        const opts = options || {};
        for (const key in opts) {
            if (opts.hasOwnProperty(key)) {
                this.optionsForComponent[key] = opts[key];
            }
        }

        if (tpl && CommonUtils.isDefaultExport(tpl)) {
            // @ts-ignore
            this.optionsForComponent.resolvedTemplate = tpl.default;
            return;
        }
        this.optionsForComponent.resolvedTemplate = tpl;
    }

    static getOptionTypes(): Record<string, unknown> {
        return {
            templateName: descriptor(String).required(),
        };
    }
}

/**
 * @name UICore/Async:Async#content
 * @cfg {String} Содержимое контейнера.
 */

/**
 * @name UICore/Async:Async#templateName
 * @cfg {String} Имя асинхронно загружаемого контрола.
 * Можно использовать только {@link /doc/platform/developmentapl/interface-development/pattern-and-practice/javascript-libraries/#_2 публичные пути библиотеки}.
 */

/**
 * @name UICore/Async:Async#templateOptions
 * @cfg {Object} Параметры содержимого контейнера Async.
 */
