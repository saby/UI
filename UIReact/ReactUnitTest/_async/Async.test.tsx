import { assert } from 'chai';
import { createSandbox, stub } from 'sinon';
import { render } from 'react-dom';
import { act } from 'react-dom/test-utils';
// FIXME: типы для jsdom нигде не подцеплены, подцепим после переезда на jest
// tslint:disable-next-line:ban-ts-ignore
// @ts-ignore
import { JSDOM } from 'jsdom';

import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import { IoC, constants } from 'Env/Env';
import { default as Async }  from 'ReactUnitTest/_async/Async';

// предзагрузим модули, на которых будем тестировать Async - в тестах главное протестировать факт вызова require
import TestControlSync = require('ReactUnitTest/_async/TestControlSync');
import TestControlAsync = require('ReactUnitTest/_async/TestControlAsync');
import TestLibraryAsync = require('ReactUnitTest/_async/TestLibraryAsync');


const isBrowser = typeof window !== 'undefined';
const describeIf = (condition) => condition ? describe : describe.skip;

describeIf(isBrowser)('UICore/Async:Async', () => {
    const warns = [];
    const originalLogger = IoC.resolve('ILogger');
    let container: HTMLDivElement;
    let sandbox;
    let clock;
    let notifyLoadStub;

    function getErrorText(moduleName: string): string {
        return `Ошибка загрузки контрола "${moduleName}"\n`
            + 'Возможны следующие причины:\n\t                   • '
            + 'Ошибка в самом контроле\n\t                   • '
            + 'Долго отвечал БЛ метод в _beforeUpdate\n\t                   • '
            + 'Контрола не существует';
    }

    /**
     * Эта функция существует для того, чтобы не забывать оборачивать тики в act.
     * Это нужно, чтобы реакт реагировал на изменение времени и обновлял компоненты.
     * @param duration Значение, на которое должно продвинуться время.
     */
    function tick(duration: number): void {
        act(() => {
            clock.tick(duration);
        });
    }

    /**
     * Асинхронный аналог {@link tick}, отличается тем, что эта версия позволяет выполниться коллбекам промисов.
     * @param duration Значение, на которое должно продвинуться время.
     */
    async function tickAsync(duration: number): Promise<void> {
        return act(async () => {
            // в новой версии sinon есть clock.tickAsync, который в теории делает то же самое
            clock.tick(duration);
            await Promise.resolve();
        });
    }

    beforeEach(() => {
        // переопределяем логгер, чтобы при ошибках загрузки не упали тесты из-за сообщений логгера
        IoC.bind('ILogger', {
            warn: (message) => {
                warns.push(message);
            },
            error: originalLogger.error,
            log: originalLogger.log,
            info: originalLogger.info
        });

        sandbox = createSandbox();
        /*
        _afterMount и _afterUpdate зовутся в отдельной таске, чтобы браузер мог отрисовать кадр.
        Чтобы не делать тесты асинхронными, мы просто мокнем таймеры и сами будем управлять временем.
         */
        clock = sandbox.useFakeTimers();
        // заглушка для нотификации событий контрола, просто проверим факт вызова метода публикации события
        notifyLoadStub = sandbox.stub(Async.prototype, '_notify').withArgs('load');

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    afterEach(() => {
        IoC.bind('ILogger', originalLogger);

        clock.restore();
        sandbox.restore();

        container.remove();
    });

    // тесты поведения на сервере
    if (typeof window === 'undefined') {
        it('Синхронная загрузка контрола на сервере', () => {
            const moduleName = 'ReactUnitTest/_async/TestControlSync';

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });

            assert.isNotOk(instance.getError(), 'Поле с ошибкой должно быть пустым.');
            assert.equal(instance.getCurrentTemplateName(), moduleName);
            assert.equal(instance.getOptionsForComponent().resolvedTemplate, TestControlSync);
        });

        it('Синхронная загрузка контрола на сервере, с ошибкой', () => {
            const moduleName = 'ReactUnitTest/_async/Fail/TestControlSync';

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });

            assert.equal(instance.getError(), getErrorText(moduleName));
            assert.strictEqual(instance.getOptionsForComponent().resolvedTemplate, undefined);
        });
    }

    describe('Проверка работы Async в браузере', () => {
        let isBrowserPlatformStub;
        let compatStub;

        before(() => {
            isBrowserPlatformStub = stub(constants, 'isBrowserPlatform').value(true);
            compatStub = stub(constants, 'compat').value(false);
        });

        after(() => {
            isBrowserPlatformStub.restore();
            compatStub.restore();
        });

        it('Синхронная загрузка контрола', () => {
            const moduleName = 'ReactUnitTest/_async/TestControlSync';

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });

            assert.isNotOk(instance.getError(), 'Поле с ошибкой должно быть пустым.');
            assert.equal(instance.getCurrentTemplateName(), moduleName);
            assert.strictEqual(instance.getOptionsForComponent().resolvedTemplate, TestControlSync);
            sandbox.assert.calledOnce(notifyLoadStub);  // Должен быть вызван метод публикации события "_notify"
        });

        it('Синхронная загрузка контрола, с ошибкой', () => {
            const moduleName = 'ReactUnitTest/_async/Fail/TestControlSync';

            // заглушка для проверки факта вызова загрузки в "require"
            const loadSyncStub = sandbox.stub(ModulesLoader, 'loadSync')
                .withArgs(moduleName).returns(null);
            // заглушка для ModulesLoader.isLoaded, чтобы проверить синхронную загрузку контрола
            const isLoadedStub = sandbox.stub(ModulesLoader, 'isLoaded');
            isLoadedStub.withArgs(moduleName).returns(true);

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });
            // удалим заглушку функции ModulesLoader.isLoaded
            isLoadedStub.restore();

            // проверим, что была попытка синхронной загрузки контрола
            sandbox.assert.called(loadSyncStub);
            assert.equal(instance.getError(), getErrorText(moduleName));
            assert.strictEqual(instance.getOptionsForComponent().resolvedTemplate, undefined);
        });

        it('Асинхронная загрузка контрола', (done) => {
            const moduleName = 'ReactUnitTest/_async/TestControlAsync';
            // заглушка для проверки факта вызова загрузки в "require"
            const loadAsyncStub = sandbox.stub(ModulesLoader, 'loadAsync')
                .withArgs(moduleName).resolves(TestControlAsync);
            // заглушка для ModulesLoader.isLoaded, чтобы проверить асинхронную загрузку контрола
            const isLoadedStub = sandbox.stub(ModulesLoader, 'isLoaded');
            isLoadedStub.withArgs(moduleName).returns(false);

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });
            // удалим заглушку функции ModulesLoader.isLoaded
            isLoadedStub.restore();

            tick(0);  // вызов _afterMount

            // проверим, что был вызов асинхронной загрузки контрола
            sandbox.assert.called(loadAsyncStub);

            // вызов _beforeUpdate
            tickAsync(0).then(() => {
                tick(0);  // вызов _afterUpdate, чтобы опубликовалось событие 'load'

                assert.isNotOk(instance.getError(), 'Поле с ошибкой должно быть пустым.');
                assert.equal(instance.getCurrentTemplateName(), moduleName);
                assert.strictEqual(instance.getOptionsForComponent().resolvedTemplate, TestControlAsync);
                sandbox.assert.calledOnce(notifyLoadStub);  // Должен быть вызван метод публикации события "_notify"
                done();
            }).catch((err) => {
                done(err);
            });
        });

        it('Асинхронная загрузка контрола, с ошибкой', (done) => {
            const moduleName = 'ReactUnitTest/_async/Fail/TestControlAsync';
            // заглушка для проверки факта вызова загрузки в "require"
            const loadAsyncStub = sandbox.stub(ModulesLoader, 'loadAsync').withArgs(moduleName).rejects();

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });

            tick(0);  // вызов _afterMount

            // проверим, что был вызов асинхронной загрузки контрола
            sandbox.assert.called(loadAsyncStub);

            // вызов _beforeUpdate
            tickAsync(0).then(() => {
                tick(0);  // вызов _afterUpdate

                assert.equal(instance.getError(), getErrorText(moduleName));
                assert.strictEqual(instance.getOptionsForComponent().resolvedTemplate, undefined);
                sandbox.assert.notCalled(notifyLoadStub);  // не должно быть вызова метода публикации события "_notify"
                done();
            }).catch((err) => {
                done(err);
            });
        });

        it('Асинхронная загрузка из библиотеки', (done) => {
            const moduleName = 'ReactUnitTest/_async/TestLibraryAsync:ExportControl';
            // заглушка для проверки факта вызова загрузки в "require"
            sandbox.stub(ModulesLoader, 'loadAsync').withArgs(moduleName).resolves(TestLibraryAsync.ExportControl);
            // заглушка для ModulesLoader.isLoaded, чтобы проверить асинхронную загрузку контрола
            const isLoadedStub = sandbox.stub(ModulesLoader, 'isLoaded');
            isLoadedStub.withArgs(moduleName).returns(false);

            let instance;
            act(() => {
                instance = render(
                    <Async templateName={moduleName} templateOptions="{}"/>,
                    container
                );
            });
            // удалим заглушку функции ModulesLoader.isLoaded
            isLoadedStub.restore();

            tick(0);  // вызов _afterMount

            // вызов _beforeUpdate
            tickAsync(0).then(() => {
                tick(0);  // вызов _afterUpdate

                assert.isNotOk(instance.getError(), 'Поле с ошибкой должно быть пустым.');
                assert.equal(instance.getCurrentTemplateName(), moduleName);
                assert.strictEqual(instance.getOptionsForComponent().resolvedTemplate, TestLibraryAsync.ExportControl);
                done();
            }).catch((err) => {
                done(err);
            });
        });
    });
});
