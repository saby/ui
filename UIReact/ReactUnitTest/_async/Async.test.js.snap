// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UICore/Async:Async на СП Синхронная загрузка контрола на сервере 1`] = `"<div tabindex=\\"0\\"></div>"`;

exports[`UICore/Async:Async на СП Синхронная загрузка контрола на сервере, с ошибкой: 1. ошибка 1`] = `"Couldn't load module \\"ReactUnitTest/_async/Fail/TestControlSync\\""`;

exports[`UICore/Async:Async на СП Синхронная загрузка контрола на сервере, с ошибкой: 2. верстка 1`] = `"<div tabindex=\\"0\\">У СБИС возникла проблема</div>"`;
